### PR TITLE
Zero downtime ECS deployments when a subnet has been modified

### PR DESCRIPTION
**Pull request by emileswarts**
# What

Call ECS API to start a rolling deployment of the current tasks, when a subnet has changed.

# Why

This allows automated deployments of configuration updates without the need to reboot the tasks manually

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -17,7 +17,9 @@ ENV LANG='C.UTF-8' \
   DB_PASS=${DB_PASS} \
   SECRET_KEY_BASE=${SECRET_KEY_BASE} \
   KEA_CONFIG_BUCKET='testbucket' \
-  DB_NAME=${DB_NAME}
+  DB_NAME=${DB_NAME} \
+  DHCP_SERVICE_NAME="some-service" \
+  DHCP_CLUSTER_NAME="some-cluster"
```
Comment (Themitchell) [inline Dockerfile:22]: Are these overridden in the ECS task definitions in prod like environments. If so perhaps a better way to add defaults for development would be to add them to the .env files.

Comment (emileswarts) [reply Dockerfile:22]: Nice yeah good idea, will add these to the .env files

**Review by jivdhaliwal — APPROVED** [review]

File: app/controllers/subnets_controller.rb [hunk 2]
```diff
@@ -72,4 +75,14 @@ def publish_kea_config
       generate_config: UseCases::GenerateKeaConfig.new(subnets: Subnet.all)
     ).execute
   end
+
+  def deploy_dhcp_service
+    UseCases::DeployKeaConfig.new(
```
Comment (jivdhaliwal) [inline app/controllers/subnets_controller.rb:80]: Should this usecase be called DeployDhcpService  instead (based on the method name and the logic in this usecase)?

Then we could add a DeployKeaConfig usecase that calls PublishKeaConfig followed by DeployDhcpService to group them together

Comment (emileswarts) [reply app/controllers/subnets_controller.rb:80]: Yeah good point. Just thinking about this, should we go with "DeployService" instead to have it general enough and reusable when we come to implementing DNS deployments?

Comment (jivdhaliwal) [reply app/controllers/subnets_controller.rb:80]: Would we be triggering a deploy of the same service to reload DNS configs ? If so then ye more generic sounds good

Comment (emileswarts) [reply app/controllers/subnets_controller.rb:80]: We will re-use this same code yeah

File: app/lib/gateways/ecs.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    end
+
+    def update_service
+      client.update_service(
+        cluster: cluster_name,
+        service: service_name,
+        force_new_deployment: true
+      )
+
+      {}
```
Comment (Themitchell) [inline app/lib/gateways/ecs.rb:18]: What s the reason for returning an empty object in this scenario? Is this used elsewhere?

File: app/controllers/subnets_controller.rb [hunk 4]
```diff
@@ -72,4 +75,14 @@ def publish_kea_config
...
   end
+
+  def deploy_service
+    UseCases::DeployService.new(
+      ecs_gateway: Gateways::Ecs.new(
+        cluster_name: ENV.fetch("DHCP_CLUSTER_NAME"),
+        service_name: ENV.fetch("DHCP_SERVICE_NAME"),
+        aws_config: Rails.application.config.s3_aws_config
+      )
+    )
```
Comment (Themitchell) [inline app/controllers/subnets_controller.rb:86]: Do we need to `execute` this use case?

File: app/lib/gateways/ecs.rb (hunk 3 above)
Comment (emileswarts) [reply app/lib/gateways/ecs.rb:18]: The default response is a large SDK response struct which isn't useful, so we noop this with an empty hash and assert that it was returned.

File: app/controllers/subnets_controller.rb (hunk 4 above)
Comment (emileswarts) [reply app/controllers/subnets_controller.rb]: Yeah some side effects like the call to ECS are difficult to observe from an acceptance test. 
If we were to go down the Webmock or VCR route, we would run into authentication errors.
Clean architecture aims to test the components in isolation and the composition of these at the call site isn't always easy to test.

Happy to discuss if you have any ideas around addressing this.

File: app/lib/gateways/ecs.rb (hunk 3 above)
Comment (Themitchell) [reply app/lib/gateways/ecs.rb:18]: 👍

**Review by Themitchell — APPROVED** [review]
Looks good. Thanks for the updates!

**Pull request merged**